metricbeat/mb/registry.go

```diff
@@ -222,16 +236,23 @@ func (r *Register) metricSetRegistration(module, name string) (MetricSetRegistra
 	name = strings.ToLower(name)
 
 	metricSets, exists := r.metricSets[module]
-	if !exists {
-		return MetricSetRegistration{}, fmt.Errorf("metricset '%s/%s' is not registered, module not found", module, name)
+	if exists {
+		registration, exists := metricSets[name]
+		if exists {
+			return registration, nil
+		}
 	}
 
-	registration, exists := metricSets[name]
-	if !exists {
-		return MetricSetRegistration{}, fmt.Errorf("metricset '%s/%s' is not registered, metricset not found", module, name)
+	// Fallback to secondary source if module is not registered
+	if source := r.secondarySource; source != nil && source.HasMetricSet(module, name) {
```

Comment:
> nit: prefer to fail early. The 'good' path return should be the last return:

I usually agree with this, but in this case there were lots of repeated lines with the same module-not-found error, so I decided to leave it as the final case. I don't have a strong preference, but for this case I think it is better like it is now.

> Question: Just trying to reduce the size of an interface, but do we need to have 'HasMetricSet' on the secondary one?

The `source.MetricSetRegistration` can return an error also if it fails to load a module, originally I made this method to return `(MetricSetRegistration, bool, error)`, and I used the bool to know if the metricset exist, but after some feedback I separated this in two methods, one to check if the module exist, and the other one to create the registration from it.

> Shouldn't 'MetricSetRegistration' error if the metricset does not exist? :)

It does, in the registry :)

Reply:
:)

I guess this is what you get for using fmt.Errorf and Wrapf for reporting errors, instead of having proper error types that can also signal the actual cause. In general I'm reluctant to add methods to an interface, but +1 for not writing `(MetricSetRegistration, bool, error)`.